src/query/service/src/sql/planner/optimizer/cascades/scheduler.rs

```diff
@@ -0,0 +1,49 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::VecDeque;
+
+use common_exception::Result;
+
+use super::tasks::Task;
+use super::CascadesOptimizer;
+
+pub struct Scheduler {
+    task_queue: VecDeque<Task>,
+}
+
+impl Scheduler {
+    pub fn new() -> Self {
+        Self {
+            task_queue: Default::default(),
+        }
+    }
+
+    pub fn run(&mut self, optimizer: &mut CascadesOptimizer) -> Result<()> {
+        while let Some(task) = self.task_queue.pop_front() {
+            if task.ref_count() > 0 {
+                // The task is still referenced by other tasks, requeue it.
+                self.task_queue.push_back(task);
+                continue;
+            }
+            task.execute(optimizer, self)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn add_task(&mut self, task: Task) {
+        self.task_queue.push_back(task);
```

Comment:
Cascades and Columbia use stack?
why use a queue here?
I thought the search progress is depth-first with a stack, so we get `final` cost of some part of the tree earlier, which can be used as Upperbound later .

Reply:
also no need ref count if using a stack?

Reply:
> why use a queue here? Cascades and Columbia use stack. I thought the search progress is depth-first with a stack, so we get `final` cost of some part of the tree earlier, which can be used as Upperbound later .

It's not actually a depth-first fashion, the task on a subtree can be terminated by another task on another subtree. 

I think there isn't much difference between queue and stack here. I would rather implement this with a recursion approach than using a state machine with the stack.

Reply:
>  I think there isn't much difference between queue and stack here
so what`s the reason to prefer queue here?

I guess stack make pruning happen earlier,  especially coupled with rule promise.

Reply:
> Task objects offer significantly more flexibility than procedure invocations, in particu- lar with respect to search algorithm and search control. A task object exists for each task that has yet to be done; all such task objects are collected in a task structure. The task structure is currently realized as a last-in-first-out stack; however, other structures can easily be envisioned. In particular, task objects can be reordered very easily at any point, enabling very flexible mechanisms for heuristic guidance. Moreover, we plan on representing the task structure by a graph that captures dependencies or the topological ordering among tasks and permit efficient parallel search (using shared memory). However, in order to obtain a working system fast, the current implemen- tation is restricted to a LIFO stack, and scheduling a task is very similar to invoking a function, with the exception that any work to be done after a sub-task completes must be scheduled as a separate task